File: Examples/intel_examples/examples_core_f/dftf/source/config_dump_descriptor.f90
!===============================================================================
! Copyright 2011-2020 Intel Corporation.
!
! This software and the related documents are Intel copyrighted  materials,  and
! your use of  them is  governed by the  express license  under which  they were
! provided to you (License).  Unless the License provides otherwise, you may not
! use, modify, copy, publish, distribute,  disclose or transmit this software or
! the related documents without Intel's prior written permission.
!
! This software and the related documents  are provided as  is,  with no express
! or implied  warranties,  other  than those  that are  expressly stated  in the
! License.
!===============================================================================

! Content:
!      Example of printing DFTI descriptor's configuration
!
!*****************************************************************************

program config_dump_descriptor

  use MKL_DFTI, forget => DFTI_DOUBLE, DFTI_DOUBLE => DFTI_DOUBLE_R

  ! Execution status
  integer :: status = 0, ignored_status

  ! DFTI descriptor handle
  type(DFTI_DESCRIPTOR), POINTER :: hand

  hand => null()

  print *,"Example config_dump_descriptor"

  print *,"Create a DFTI descriptor"
  status = DftiCreateDescriptor(hand, DFTI_DOUBLE, DFTI_REAL, 3, [5,6,7])
  if (0 /= status) goto 999

  print *,"Dump the descriptor"
  call dump_descriptor(hand)

100 continue

  print *,"Free the descriptor"
  ignored_status = DftiFreeDescriptor(hand)

  if (status == 0) then
    print *, "TEST PASSED"
    call exit(0)
  else
    print *, "TEST FAILED"
    call exit(1)
  end if

999 continue
  print '("  Error, status = ",I0)', status
  goto 100

contains

  subroutine dump_descriptor(hand)
    integer, parameter :: MR = 10 ! maximum rank
    type(DFTI_DESCRIPTOR), POINTER :: hand
    integer :: prec, domain, rank, lengths(MR)
    integer :: placement
    real*8  :: fwd_scale, bwd_scale
    integer :: nut, istrides(1+MR), ostrides(1+MR), ntr, idist, odist
    integer :: storage, packfmt, wspace, cmtstatus

    ! Execution status
    integer :: status

    print '("  PRECISION              = "$)'
    status = DftiGetValue(hand, DFTI_PRECISION, prec)
    if (0 /= status) goto 999

    if      (prec == DFTI_SINGLE) then ; print '("DFTI_SINGLE")'
    else if (prec == DFTI_DOUBLE) then ; print '("DFTI_DOUBLE")'
    else                               ; goto 999
    end if

    print '("  FORWARD_DOMAIN         = "$)'
    status = DftiGetValue(hand, DFTI_FORWARD_DOMAIN, domain)
    if (0 /= status) goto 999

    if      (domain == DFTI_COMPLEX) then ; print '("DFTI_COMPLEX")'
    else if (domain == DFTI_REAL)    then ; print '("DFTI_REAL")'
    else                                  ; goto 999
    end if

    print '("  DIMENSION              = "$)'
    status = DftiGetValue(hand, DFTI_DIMENSION, rank)
    if (0 /= status) goto 999
    print '(I0)', rank

    print '("  LENGTHS                = "$)'
    status = DftiGetValue(hand, DFTI_LENGTHS, lengths)
    if (0 /= status) goto 999
    print '(10(I0:", "))', lengths(1:rank)

    print '("  PLACEMENT              = "$)'
    status = DftiGetValue(hand, DFTI_PLACEMENT, placement)
    if (0 /= status) goto 999

    if      (placement == DFTI_INPLACE)     then ; print '("DFTI_INPLACE")'
    else if (placement == DFTI_NOT_INPLACE) then ; print '("DFTI_NOT_INPLACE")'
    else                                         ; goto 999
    end if

    print '("  F/B SCALES             = "$)'
    status = DftiGetValue(hand, DFTI_FORWARD_SCALE,  fwd_scale)
    if (0 /= status) goto 999
    status = DftiGetValue(hand, DFTI_BACKWARD_SCALE, bwd_scale)
    if (0 /= status) goto 999
    print '(2(G11.4:", "))', fwd_scale, bwd_scale

    print '("  NO OF USER THREADS     = "$)'
    status = DftiGetValue(hand, DFTI_NUMBER_OF_USER_THREADS, nut)
    if (0 /= status) goto 999
    print '(I0)', nut

    print '("  INPUT STRIDES          = "$)'
    status = DftiGetValue(hand, DFTI_INPUT_STRIDES,  istrides)
    if (0 /= status) goto 999
    print '(I0"; "10(I0:", "))', istrides(1:1+rank)

    print '("  OUTPUT STRIDES         = "$)'
    status = DftiGetValue(hand, DFTI_OUTPUT_STRIDES, ostrides)
    if (0 /= status) goto 999
    print '(I0"; "10(I0:", "))', ostrides(1:1+rank)

    print '("  NO OF TRANSFORMS       = "$)'
    status = DftiGetValue(hand, DFTI_NUMBER_OF_TRANSFORMS, ntr)
    if (0 /= status) goto 999
    print '(I0)', ntr

    print '("  I/O DISTANCES          = "$)'
    status = DftiGetValue(hand, DFTI_INPUT_DISTANCE,  idist)
    if (0 /= status) goto 999
    status = DftiGetValue(hand, DFTI_OUTPUT_DISTANCE, odist)
    if (0 /= status) goto 999
    print '(2(I0:", "))', idist, odist

    if (domain == DFTI_COMPLEX) then
      print '("  COMPLEX STORAGE        = "$)'
      status = DftiGetValue(hand, DFTI_COMPLEX_STORAGE, storage)
      if (0 /= status) goto 999
      if (storage == DFTI_COMPLEX_COMPLEX) then
        print '("DFTI_COMPLEX_COMPLEX")'
      else if (storage == DFTI_REAL_REAL) then
        print '("DFTI_REAL_REAL")'
      else
        goto 999
      end if
    else if (domain == DFTI_REAL) then
      print '("  CONJUGATE EVEN STORAGE = "$)'
      status = DftiGetValue(hand, DFTI_CONJUGATE_EVEN_STORAGE, storage)
      if (0 /= status) goto 999
      if (storage == DFTI_COMPLEX_COMPLEX) then
        print '("DFTI_COMPLEX_COMPLEX")'
      else if (storage == DFTI_COMPLEX_REAL) then
        print '("DFTI_COMPLEX_REAL")'
      else
        goto 999
      end if
      if (storage == DFTI_COMPLEX_REAL) then
        print '("    PACKED FORMAT        = "$)'
        status = DftiGetValue(hand, DFTI_PACKED_FORMAT, packfmt)
        if (0 /= status) goto 999
        if (packfmt == DFTI_CCS_FORMAT) then
          print '("DFTI_CCS_FORMAT")'
        else if (packfmt == DFTI_PACK_FORMAT) then
          print '("DFTI_PACK_FORMAT")'
        else if (packfmt == DFTI_PERM_FORMAT) then
          print '("DFTI_PERM_FORMAT")'
        else
          goto 999
        end if
      end if
    end if

    print '("  WORKSPACE              = "$)'
    status = DftiGetValue(hand, DFTI_WORKSPACE, wspace)
    if (0 /= status) goto 999

    if      (wspace == DFTI_ALLOW) then ; print '("DFTI_ALLOW")'
    else if (wspace == DFTI_NONE ) then ; print '("DFTI_AVOID")'
    else                                ; goto 999
    end if

    print '("  COMMIT STATUS          = "$)'
    status = DftiGetValue(hand, DFTI_COMMIT_STATUS, cmtstatus)
    if (0 /= status) goto 999

    if (cmtstatus == DFTI_COMMITTED) then
      print '("DFTI_COMMITTED")'
    else if (cmtstatus == DFTI_UNCOMMITTED) then
      print '("DFTI_UNCOMMITTED")'
    else
      goto 999
    end if

    return

999 print *," Error, status = ", status
    call exit(1)

  end subroutine dump_descriptor

end program config_dump_descriptor

File: Examples/intel_examples/examples_core_c/dftc/source/config_dump_descriptor.c
/*******************************************************************************
* Copyright 2011-2020 Intel Corporation.
*
* This software and the related documents are Intel copyrighted  materials,  and
* your use of  them is  governed by the  express license  under which  they were
* provided to you (License).  Unless the License provides otherwise, you may not
* use, modify, copy, publish, distribute,  disclose or transmit this software or
* the related documents without Intel's prior written permission.
*
* This software and the related documents  are provided as  is,  with no express
* or implied  warranties,  other  than those  that are  expressly stated  in the
* License.
*******************************************************************************/

/*
! Content:
! Example of printing DFTI descriptor's configuration
!
!****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include "mkl_dfti.h"

static void dump_descriptor(DFTI_DESCRIPTOR_HANDLE hand);

/* Define the format to printf MKL_LONG values */
#if !defined(MKL_ILP64)
#define LI "%li"
#else
#define LI "%lli"
#endif

int main(void)
{
    /* Sizes of a 4D FFT */
    MKL_LONG N[4] = {10, 20, 30, 40};

    /* Execution status */
    MKL_LONG status = 0;

    DFTI_DESCRIPTOR_HANDLE hand = NULL;

    printf("Create a DFTI descriptor\n");
    status = DftiCreateDescriptor(&hand, DFTI_DOUBLE, DFTI_COMPLEX, 4, N);
    if (status != DFTI_NO_ERROR) goto failed;

    printf("Dump the descriptor\n");
    dump_descriptor(hand);

 cleanup:

    printf("Free DFTI descriptor\n");
    DftiFreeDescriptor(&hand);

    printf("TEST %s\n", (status == 0) ? "PASSED" : "FAILED");
    return status;

 failed:
    printf(" ERROR, status = "LI"\n", status);
    status = 1;
    goto cleanup;
}

/* Maximum supported rank of multidimensional FFTs */
#define MAX_RANK 7

static void dump_descriptor(DFTI_DESCRIPTOR_HANDLE hand)
{
    /* Execution status */
    MKL_LONG status = 0;

    char version[DFTI_VERSION_LENGTH];

    enum DFTI_CONFIG_VALUE placement, precision, domain, storage, packfmt,
                           wspace, cmtstatus;
    MKL_LONG rank, lengths[MAX_RANK];
    double fwd_scale, bwd_scale;
    MKL_LONG nut, is[1+MAX_RANK], os[1+MAX_RANK], ntr, idist, odist, tlimit;

    DftiGetValue(0, DFTI_VERSION, version);
    printf("%s\n", version);

    printf("  PRECISION = "); fflush(0);
    status = DftiGetValue(hand, DFTI_PRECISION, &precision);
    if (status != DFTI_NO_ERROR) goto failed;
    if      (precision == DFTI_SINGLE) printf("DFTI_SINGLE\n");
    else if (precision == DFTI_DOUBLE) printf("DFTI_DOUBLE\n");
    else { printf("unknown (%i)\n", precision); goto failed; }

    printf("  FORWARD_DOMAIN = "); fflush(0);
    status = DftiGetValue(hand, DFTI_FORWARD_DOMAIN, &domain);
    if (status != DFTI_NO_ERROR) goto failed;
    if      (domain == DFTI_COMPLEX) printf("DFTI_COMPLEX\n");
    else if (domain == DFTI_REAL)    printf("DFTI_REAL\n");
    else { printf("unknown (%i)\n", domain); goto failed; }

    printf("  DIMENSION = "); fflush(0);
    status = DftiGetValue(hand, DFTI_DIMENSION, &rank);
    if (status != DFTI_NO_ERROR) goto failed;
    printf(LI"\n", rank);

    printf("  LENGTHS = "); fflush(0);
    status = DftiGetValue(hand, DFTI_LENGTHS, lengths);
    if (status != DFTI_NO_ERROR) goto failed;

    {
        int r = 0;
        printf(LI, lengths[0]);
        for (r = 1; r < rank; ++r) printf(", "LI, lengths[r]);
        printf("\n");
    }

    printf("  PLACEMENT = "); fflush(0);
    status = DftiGetValue(hand, DFTI_PLACEMENT, &placement);
    if (status != DFTI_NO_ERROR) goto failed;
    if      (placement == DFTI_NOT_INPLACE) printf("DFTI_NOT_INPLACE\n");
    else if (placement == DFTI_INPLACE)     printf("DFTI_INPLACE\n");
    else { printf("unknown (%i)\n", placement); goto failed; }

    printf("  F/B SCALES = "); fflush(0);
    if (precision == DFTI_DOUBLE) {
        status = DftiGetValue(hand, DFTI_FORWARD_SCALE, &fwd_scale);
        if (status != DFTI_NO_ERROR) goto failed;
        status = DftiGetValue(hand, DFTI_BACKWARD_SCALE, &bwd_scale);
        if (status != DFTI_NO_ERROR) goto failed;
    } else {
        float fs, bs;
        status = DftiGetValue(hand, DFTI_FORWARD_SCALE, &fs);
        if (status != DFTI_NO_ERROR) goto failed;
        status = DftiGetValue(hand, DFTI_BACKWARD_SCALE, &bs);
        if (status != DFTI_NO_ERROR) goto failed;
        fwd_scale = (double)fs;
        bwd_scale = (double)bs;
    }
    printf(" %lg, %lg\n", fwd_scale, bwd_scale);

    printf("  NO OF USER THREADS = "); fflush(0);
    status = DftiGetValue(hand, DFTI_NUMBER_OF_USER_THREADS, &nut);
    if (status != DFTI_NO_ERROR) goto failed;
    printf(LI"\n", nut);

    printf("  INPUT  STRIDES = "); fflush(0);
    status = DftiGetValue(hand, DFTI_INPUT_STRIDES, is);
    if (status != DFTI_NO_ERROR) goto failed;

    {
        int r = 0;
        printf(LI, is[0]);
        for (r = 1; r <= rank; ++r) printf(", "LI, is[r]);
        printf("\n");
    }

    printf("  OUTPUT STRIDES = "); fflush(0);
    status = DftiGetValue(hand, DFTI_OUTPUT_STRIDES, os);
    if (status != DFTI_NO_ERROR) goto failed;

    {
        int r = 0;
        printf(LI, os[0]);
        for (r = 1; r <= rank; ++r) printf(", "LI, os[r]);
        printf("\n");
    }

    printf("  NO OF TRANSFORMS = "); fflush(0);
    status = DftiGetValue(hand, DFTI_NUMBER_OF_TRANSFORMS, &ntr);
    if (status != DFTI_NO_ERROR) goto failed;
    printf(LI"\n", ntr);

    printf("  I/O DISTANCES = "); fflush(0);
    status = DftiGetValue(hand, DFTI_INPUT_DISTANCE, &idist);
    if (status != DFTI_NO_ERROR) goto failed;
    status = DftiGetValue(hand, DFTI_OUTPUT_DISTANCE, &odist);
    if (status != DFTI_NO_ERROR) goto failed;
    printf(LI", "LI"\n", idist, odist);

    if (domain == DFTI_COMPLEX) {
        printf("  COMPLEX STORAGE = "); fflush(0);
        status = DftiGetValue(hand, DFTI_COMPLEX_STORAGE, &storage);
        if (status != DFTI_NO_ERROR) goto failed;
        if (storage == DFTI_COMPLEX_COMPLEX) printf("DFTI_COMPLEX_COMPLEX\n");
        else if (storage == DFTI_REAL_REAL)  printf("DFTI_REAL_REAL\n");
        else { printf("wrong (%i)\n", storage); goto failed; }
    } else {
        printf("  CONJUGATE EVEN STORAGE = "); fflush(0);
        status = DftiGetValue(hand, DFTI_CONJUGATE_EVEN_STORAGE, &storage);
        if (status != DFTI_NO_ERROR) goto failed;
        if (storage == DFTI_COMPLEX_COMPLEX)   printf("DFTI_COMPLEX_COMPLEX\n");
        else if (storage == DFTI_COMPLEX_REAL) printf("DFTI_COMPLEX_REAL\n");
        else { printf("wrong (%i)\n", storage); goto failed; }
        if (storage == DFTI_COMPLEX_REAL) {
            printf("     PACKED FORMAT = "); fflush(0);
            status = DftiGetValue(hand, DFTI_PACKED_FORMAT, &packfmt);
            if (status != DFTI_NO_ERROR) goto failed;
            if      (packfmt == DFTI_CCS_FORMAT) printf("DFTI_CCS_FORMAT\n");
            else if (packfmt == DFTI_PACK_FORMAT) printf("DFTI_PACK_FORMAT\n");
            else if (packfmt == DFTI_PERM_FORMAT) printf("DFTI_PERM_FORMAT\n");
            else { printf("wrong (%i)\n", packfmt); goto failed; }
        }
    }

    printf("  WORKSPACE = "); fflush(0);
    status = DftiGetValue(hand, DFTI_WORKSPACE, &wspace);
    if (status != DFTI_NO_ERROR) goto failed;
    if      (wspace == DFTI_ALLOW) printf("DFTI_ALLOW\n");
    else if (wspace == DFTI_AVOID) printf("DFTI_AVOID\n");
    else if (wspace == DFTI_NONE)  printf("DFTI_NONE\n");
    else { printf("wrong (%i)\n", wspace); goto failed; }

    printf("  COMMIT STATUS = "); fflush(0);
    status = DftiGetValue(hand, DFTI_COMMIT_STATUS, &cmtstatus);
    if (status != DFTI_NO_ERROR) goto failed;
    if      (cmtstatus == DFTI_COMMITTED) printf("DFTI_COMMITTED\n");
    else if (cmtstatus == DFTI_UNCOMMITTED) printf("DFTI_UNCOMMITTED\n");
    else { printf("wrong (%i)\n", cmtstatus); goto failed; }

    printf("  THREAD LIMIT = "); fflush(0);
    status = DftiGetValue(hand, DFTI_THREAD_LIMIT, &tlimit);
    if (status != DFTI_NO_ERROR) goto failed;
    printf(LI"\n", tlimit); fflush(0);

    return;

 failed:
    printf("Error, status = "LI"\n", status);
    exit(1);
}
